Add shuffle article list

j605 opened the pull request:
This PR allows one to have random ordering to article list. The motivation for this was to display a set of random unread articles using query feeds although getting random sort order per-feed is not currently possible, this is a workaround. 

As far as tests are concerned, I am not sure how to do it. An idea would be have a fixed seed for `random_shuffle`, but then this has to exposed somehow from `sort_unlocked`.

j605 commented:
Regarding using `default_random_engine`, we need to use `shuffle` I think. I also found out out that `random_shuffle` is deprecated and will go away by C++17 so moving to `shuffle` should be fine. That has been available from C++11. I am clueless as to how to specify default arguments as functions, here it is `random(Iterator, Iterator, URBG&&)` since we also need to instantiate `default_random_engine`.

Minoru commented:
Yup, `shuffle` is good. I don't expect Newsboat to ever require C++14 or C++17, but it's good to pick the most available function on a principle.

> I am clueless as to how to specify default arguments as functions, here it is random(Iterator, Iterator, URBG&&) since we also need to instantiate default_random_engine.

`default_random_engine` is a function object, not a function, so you just instantiate it:
```c++
void f(std::default_random_engine&& rng = std::default_random_engine());
```

However, this won't work in our case because `URBG` is a template. Thus we can't write a non-template wrapper around `shuffle`. In other words, `sort_unlocked` has to become a template method:
```c++
// This template function will be used in test code
template <class URBG>
void sort_unlocked(
    const ArticleSortStrategy& sort_strategy,
    URBG&& rng)
{
    // the body of existing sort_unlocked here
}

// This function will be used in production code because it
// doesn't require the second argument
void sort_unlocked(const ArticleSortStrategy& sort_strategy)
{
    // std::random_device might be a better fit here since it
    // *should* be initialized with a random seed, whereas
    // default_random_generator doesn't provide such promises
    sort_unlocked(sort_strategy, std::default_random_engine());
}
```

This is ugly and I don't like this.

Let's look at the problem from a different angle: what are we going to test? If we substitute the PRNG with our own, we'll be able to check that `sort_unlocked` orders items according to the values PRNG returned. Can we just check that the sorting is random?

With 5 test items, there are 5!=120 possible orderings. If we sort once and compare to the original ordering, the chances of them being the same are 1 in 120 (assuming all 5 items are different). That's 0,83%, which is quite high. If we sort again, the chances of all three orderings to be the same are 1 in (1/120)^2, or 0,007%. In statistics, that would be considered good. Given how fast this check is, we can sort 10 times, getting an astronomically small probability of a test failure.

If we go this route, we'd better add an `INFO` to the test, saying that its failure doesn't mean the function is broken, only that it *might* be broken.

What do you think of that, @j605?

(No code in this post was tested or even compiled. And sorry for the wall of text.)

Minoru commented:
Another idea, suggested by a friend: move `sort_unlocked` into a separate template class. Its template argument is URBG. In `RssFeed`, we'll add an instance parameterized by `std::random_device` or `std::default_random_engine`; in tests, we'll use a generator with a known seed. Tests won't check `RssFeed::sort_unlocked` because we'll know that it just calls our "sorter" class.

This is flawed in its own ways, but maybe this can be refined into something better.

j605 commented:
I was leaning towards using a constant seed to have reproducible shuffle pattern which can be tested. To be honest I am not convinced as to what we want to test. If the test uses different code then how can we detect failure when the actual code is changed.

Minoru commented:
Okay then. Make `sort_unlocked` a template method and provide a wrapper that calls it with `default_random_engine` or `random_device` (probably the latter, I suspect former uses the same seed every time). In tests, pass whatever RNG you want as an argument to the template method. I'll have to give up on perfectionism here :)

j605 commented:
@Minoru I refactored and added a test that will surely fail if the test code compiles. I just get a huge wall linker errors and it is not apparent what is causing the issue. Also I had to move the function into the header file as I found out template functions cannot be virtualized.

Minoru commented on include/rss.h:
```diff
@@ -344,7 +346,98 @@ class RssFeed : public Matchable {
+		}
+	}
+
+	void sort_unlocked_wrapper(const ArticleSortStrategy& sort_strategy);
```
There is no need to change the name:
1. C++ lets us overload templated functions just like any other function.
2. in my opinion, having "wrapper" mentioned throughout the code is kinda ugly, and I'd rather avoid that.

Minoru commented:
> I just get a huge wall linker errors and it is not apparent what is causing the issue.

All Travis builds compiled fine, and my local build succeeded as well, so it's probably just some file that wasn't recompiled. Run `make distclean` and try again.

It's interesting to note GCC 7 and 8 on Linux sorted the items differently from all other compilers. I'm not exactly sure how best to deal with this, but I suggest doing the following:
* use `std::linear_congruential_engine` directly rather than relying on `std::minstd_rand0` typedef (perhaps there is a bug in seeds in GCC 7 and 8);
* check the first value returned by the engine, and if it doesn't match the expected one, fail the test with an explanation that the PRNG seem to be implemented incorrectly.

j605 commented:
Fixed the tests :)

Minoru commented:
They will only pass on GCC 7 and 8 because of RNG implementation problem I described :(

j605 commented:
Should I use `std::linear_congruential_engine<std::uint_fast32_t, 16807, 0, 2147483647>` which is the definition of `std::minstd_rand0`. Will this fix the problem with older GCC versions?

Minoru commented:
I don't know for sure (I only have GCC 8 locally), but I hope it will.

j605 commented:
This worked locally with clang, so hopefully it works on other compiler versions as well.

Minoru commented:
I don't understand what's going on over at Travis. I tried Clang 7 locally, and the test passes just fine. I'm on Debian, so my Clang should be fairly similar to that in Ubuntu.

Googling didn't provide any answers yet. @j605, you might try to use other generators in hopes that one of them will be consistent across all platforms. Don't have any better suggestions at the moment.

Minoru commented:
To clarify: I'm still researching and haven't given up on this yet. If no other standard generator works, we can implement our own (URBG is a simple object with just four functions).

Minoru commented:
std::mt19937 didn't fare any better :( So, how about implementing our own? [it's only four functions](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator), and since it's just a test, our generator can simply return pre-defined integers (even just a sequence of 0, 1, 2, ...)

j605 commented:
This seems to have worked :)

Minoru commented on test/rss.cpp:
```diff
@@ -489,6 +491,48 @@ TEST_CASE("RssFeed::sort() correctly sorts articles", "[rss]")
+			public:
+				typedef uint32_t result_type;
+
+				result_type min()
```
This and `max()` need to be `static`; MacOS [tries to call `G::min()`](https://travis-ci.org/newsboat/newsboat/jobs/500580116#L2623) and fails because this implementation of `min()` needs an object (i.e. `G::min()` is invalid, only `rng.min()` is).

Minoru commented on test/rss.cpp:
```diff
@@ -489,6 +491,48 @@ TEST_CASE("RssFeed::sort() correctly sorts articles", "[rss]")
+					return 5;
+				}
+
+				result_type operator()()
```
It took me a minute to understand why this passed all tests even though `i` is never initialized. My understanding is this:
1. `i` gets initialized with whatever garbage was in memory at the time when `i` was allocated;
2. when this operator is called, there is an extremely high probability ((2^32-6)/(2^32), 99.99999986030161%) that `i` is outside of [0; 5] range. Thus the `if` is almost always executed, `i` is initialized to 1, and the following calls behave deterministically (i.e. always the same).

Please initialize `i` to something to eliminate that 0.00000013969839% chance of failure :)

j605 replied:
Regarding `i` I thought that private members are initialized to zero or something. I should've gone with a known quantity.

j605 commented:
I think it failed because `max` and `min` were not marked as `constexpr` functions. I am uploading that change now.

j605 commented:
It failed again but this time the sort order is different in macos, this looks impossible since the URBG gives out the same sequence  everywhere (1, 2, 3, 4, 5) or is something else happening there?

Minoru commented:
It's quite possible that the standard library in macOS implemets `shuffle` very differently from Linux, e.g. calling the generator three times instead of two.

\*sigh*

Let's remove the tests for now, and merge the rest as-is. I'll open an issue to re-visit this later. I feel like we're approaching this wrong and run into issues that we shouldn't be having, but I can't tell what exactly is wrong here. Maybe I (or someone else) will figure it out later.

Sorry I've wasted so much of your time chasing ideas that all led nowhere.  But it's said that "experience is what you get when you don't get what you wanted"; we certainly gained quite a bit of experience with this PR :)

j605 commented:
Would you prefer a revert or should I just remove the tests commit and retain only the first commit. I can save the last commit in a different branch if we are going to look at it in the future.

Minoru commented:
Actually, let's not touch history at all. I'll cherry-pick the first commit, close this PR, and open an issue referencing it. No action required from you.

Should've thought of it yesterday, sorry about that.

Minoru commented:
Okay, the issue is #461. Thank you for your work here, @j605!
